### ISSUE TITLE
Firebase Test Lab Robo Test support

### ISSUE THREAD

**Issue by jaumard** [flutter/flutter#36681]
Firebase offer a nice way to test android app automatically (or semi automatically if you provide a Robo Script) called Robo test in their Test Lab section https://firebase.google.com/docs/test-lab/android/robo-ux-test

Would be nice if Flutter application can be test that way too.

Currently when you upload the APK, there is 2 issues:
- it open the app, wait for the first screen to show, then exit the app. It doesn't try to automatically explore the app
- We can't provide a Robo Script because the AS plugin to do that doesn't record anything for a flutter app

That way of testing is quite nice as it doesn't need that much work to actually do something, and firebase allow a large choice of devices. Doing such UI test will help finding UI problem (so it also benefit iOS, desktop and web platform too)

**Comment by edufolly** [comment]
Nice request!

**Comment by stemuk** [comment]
I think the issue is more on the Firebase than on the Flutter side of things. The way robo tests work is that they look for AndroidUI elements (buttons, FABs and lists etc.) and click them as long as they dont hit a dead end. 

The Firebase team would now have to modify their audtomated tests to work with Flutter UI elements too, since flutter uses custom elements only. Right now, all the Robo Tester 'sees' is a blank 'faketouch' canvas, since this is the only native element flutter uses.

**Comment by jaumard** [comment]
Yes but if flutter doesn't allow to retrieve FlutterUI elements to native side they can't do anything, also we can imagine mapping back to equivalent AndroidUI elements directly on the engine. Might be faster than waiting for the Firebase team to implement a flutter version that would also depends on the flutter team at the end... So my guess is that most of the work is here anyway. And if not flutter team can talk to firebase team about this more easily than I can. 

So enlightenment from the team would be welcome for this ^^

**Comment by vishna** [comment]
It doesn't seem like this will work with Firebase Lab unless Firebase team starts officially supporting Flutter instrumentation - since it's all Google maybe they can talk to each other. In the meantime it looks like the Amazon farm + [sylph](https://github.com/mmcc007/sylph) is the clear winner here.

**Comment by jaumard** [comment]
@vishna `Amazon farm + sylph` doesn't offer the same functionality here, as you need to write integration tests ^^ but I guess that's the only way for now...

**Comment by sceee** [comment]
I would be fine if the pre-launch-test in the first place just runs my integration tests. Random inputs could come later.

**Comment by kroikie** [comment]
@jaumard

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/353. Any further collaboration will be done there.

**Comment by collinjackson** [comment]
This isn't an issue with Flutterfire plugins, but rather the exposing the Flutter widget tree to the Robo tester, so I've closed the Flutterfire issue and reopened this one.

The [e2e package](https://pub.dev/packages/e2e) can be used to author tests that use the [flutter_test](https://api.flutter.dev/flutter/flutter_test/flutter_test-library.html) package on Firebase Test Lab and get the results in the instrumentation test format that Firebase Test Lab understands. We're using this approach to test the [plugins](https://github.com/flutter/plugins) that are maintained by the Flutter team as well as [Flutterfire](https://github.com/FirebaseExtended/flutterfire).

https://cirrus-ci.com/task/5458389978578944

https://cirrus-ci.com/task/6322880593264640

Instrumentation testing fulfills a different need than random Robo testing, but it has some advantages: tests are written in the same language as the app and have full access to the built widget tree.

**Comment by jaumard** [comment]
Thanks for the explanation @collinjackson ! 
I was wondering something, if instrumentation testing have access to the built tree, does it mean that technically we can create:

- a robot that will do dumb test like the one in firebase labs ? 
- a tool that will register user action to save script that can be replayed later ? 

Let me know :) can be interesting to do !

**Comment by sceee** [comment]
@collinjackson I would be totally happy if the Testlab tests execute one/multiple of my _defined (handwritten)_ tests instead of random input tests.
But how will Google Play detect which of the tests to execute as a replacement of the default random test it executes?

**Comment by TatsuUkraine** [comment]
Hello, wanted to ask is there any timeline or plans to add support robo script recording? I see that this issue doesn't have any activity since 2023, so really appreciate for any update. Thanks!
